interfaces/builtin/appstream_metadata.go

```diff
@@ -0,0 +1,120 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/mount"
+	"github.com/snapcore/snapd/osutil"
+)
+
+const appstreamMetadataSummary = `allows access to AppStream metadata`
+
+const appstreamMetadataBaseDeclarationSlots = `
+  appstream-metadata:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+// Paths for upstream and collection metadata are defined in the
+// AppStream specification:
+//   https://www.freedesktop.org/software/appstream/docs/
+const appstreamMetadataConnectedPlugAppArmor = `
+# Description: Allow access to AppStream metadata from the host system
+
+# Allow access to AppStream upstream metadata files
+/usr/share/metainfo/** r,
+/usr/share/appdata/** r,
+
+# Allow access to AppStream collection metadata
+/usr/share/app-info/** r,
+/var/cache/app-info/** r,
+/var/lib/app-info/** r,
+
+# Apt symlinks the DEP-11 metadata to files in /var/lib/apt/lists
+/var/lib/apt/lists/*.yml.gz r,
+`
+
+var appstreamMetadataDirs = []string{
```

Comment:
I'm somewhat worried about the number of mount entries this requires. This is will impact people's ability to use layouts.

Reply:
I don't expect a huge number of snaps to end up using this interface: it is only going to be of interest to package management tools, where the pros of access to the data outweigh the cons.

I had considered using `system-files` to punch holes to paths in `/var/lib/snapd/hostfs/...` as an alternative, but the mounts have a few benefits:

1. both libraries for reading AppStream metadata contain hard coded paths in the code:

    https://github.com/ximion/appstream/blob/master/src/as-pool.c#L95
    https://github.com/hughsie/appstream-glib/blob/master/libappstream-glib/as-store.c#L3326

2. On Ubuntu and Debian systems, the collection metadata files in `/var/lib/app-info/yaml` are absolute symlinks that won't resolve using the hostfs paths.

Reply:
> I'm somewhat worried about the number of mount entries this requires. This is will impact people's ability to use layouts.

you mean the potential for conflicts or just global limits on the number mounts?

Reply:
Potential conflicts, we don't currently enforce any numeric limits.

Reply:
> I'm somewhat worried about the number of mount entries this requires. This is will impact people's ability to use layouts.

It's conceivable that snapd could could check for conflicting layouts if the snap plugs this interface. This can also happen in the review-tools as an intermediate step to not block this PR. That said, IME, only the /var/lib/apt/lists is at all worrisome since the others are very clearly appstream.

One thought is that since @jhenstridge mentioned this is only probably for package management tools and we consider that type of snap to be highly trusted, we could give this interface an installation constraint and avoid the layouts issue in practice. The data itself doesn't warrant it, but we can add a code comment on why we have the installation constraint. This would not be onerous for legitimate users because they likely need packagekit-control (or some future, similarly powerful interface) which needs an installation request anyway. I'm not necessarily recommending this path, but like I said, it would avoid the layouts conflict issue in practice and could be removed at a future date when conflicts are detected.

Reply:
I'd be happy to add a plug restriction to the interface: given that the only current user will need one for `packagekit-control` anyway, it isn't a huge burden.

As far as information disclosure goes, there's two types of info exposed:

1. collection metadata: this is metadata about packages available in the distro repository.
2. upstream metadata: this is metadata provided by installed packages.

While (1) only really leaks information about configured software repositories on the host,  (2) leaks information about (some of) the installed applications.  While we definitely shouldn't allow auto-connection by default, I'm not sure if this goes all the way to denying installation by default.

Reply:
@zyga - can you comment on how concerned you are about the potential layout conflict? Personally, I don't think we should block this PR on this detection and this could come in future PRs. If you are quite concerned, we can cheaply use the installation constraint or add a review-tools check until the layout detection is in place.

Reply:
Given that this interface is to be used by specific applications only I'm no longer concerned. On my first read I didn't mentally scope it to a specific application and I assumed anyone would be interested in connecting it. With the additional information I'm +1 to merge it as well as to come up with better conflict detection, separately.